iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/event/common/tsfile/TsFileInsertionDataContainer.java

```diff
@@ -95,28 +97,46 @@ public TsFileInsertionDataContainer(
     this.sourceEvent = sourceEvent;
 
     try {
-      allocatedMemoryBlock =
-          PipeResourceManager.memory()
-              .forceAllocate(
-                  PipeConfig.getInstance().getPipeMemoryAllocateForTsFileSequenceReaderInBytes());
-
-      tsFileSequenceReader = new TsFileSequenceReader(tsFile.getAbsolutePath(), true, true);
-      tsFileReader = new TsFileReader(tsFileSequenceReader);
-
-      deviceMeasurementsMapIterator = filterDeviceMeasurementsMapByPattern().entrySet().iterator();
-      deviceIsAlignedMap = readDeviceIsAlignedMap();
-      measurementDataTypeMap = tsFileSequenceReader.getFullPathDataTypeMap();
+      tsFileSequenceReader = PipeResourceManager.tsfile().getTsFileSequenceReaderFromCache(tsFile);
+      if (tsFileSequenceReader == null) {
+        // TsFileSequenceReader is not in cache, we need to create it here and close it later.
+        isSelfHoldReader = true;
+        LOGGER.info(
+            "TsFileSequenceReader can't be created in cache, will create it outside of cache.");
+
+        allocatedMemoryBlock =
+            PipeResourceManager.memory()
+                .forceAllocate(
+                    PipeConfig.getInstance().getPipeMemoryAllocateForTsFileSequenceReaderInBytes());
+        tsFileSequenceReader = new TsFileSequenceReader(tsFile.getPath(), true, true);
+        tsFileReader = new TsFileReader(tsFileSequenceReader);
+        deviceIsAlignedMap = readDeviceIsAlignedMap();
+        measurementDataTypeMap = tsFileSequenceReader.getFullPathDataTypeMap();
+      } else {
+        tsFileReader = PipeResourceManager.tsfile().getTsFileReaderFromCache(tsFile);
```

Comment:
Is that safe to share the same tsFileReader with other threads?

Reply:
I think so, all usage I see is read-only.

Reply:
NOT SAFE. THE CODE IS MODIFIED.